### ISSUE TITLE
Print details of a search result

### ISSUE THREAD

**Issue by johan12345** [opacapp/opacclient#431]
**Hey!** If you ever wanted to contribute to an Open Source project, but didn't know how, you just found a perfect place to start. We try to write this issue as instructive as possible, but as we're humans we almost certainly forgot to write something important in here. If you want to try to tackle this issue and get stuck at *any* point, please do not hesitate and ask us, either here on GitHub or via email at info@opacapp.de. We'll be very happy to help!

We think this issue is very suitable to people who are just starting with Open Source or Android development in general and have never done it before. We'd like to ask people with wide experience in contributing to leave this issue to beginners. Thanks!

**In addition to Java/Android programming, we recommend people willing to start working on this issue to be at least somewhat familiar with HTML.**

## How to get started

First of all, we recommend that you [install the app](https://play.google.com/store/apps/details?id=de.geeksfactory.opacclient) and play around with it so you know how it looks and feels. Next, you should use git to download this GitHub repository and open the folder `opacclient/` from within the repository (not the repository itself) in Android Studio.

Because of a [Bug in Android Studio](https://code.google.com/p/android/issues/detail?id=172347), some configuration files in the `.idea` folder might be deleted after opening the project for the first time. Run ``git checkout -- opacclient/.idea/`` in the repository to revert this.

Press the green "Play" button to run it on your device. If this works, you have a working development setup, yay! If it doesn't, feel free to ask us for help!

## What this is about

When you start the app and set up a library account, you can search this library's catalog, view the search result list and click on the results to get detailed information about a single item, including things like where in the library it is available. Users who only have a tablet (not a smartphone) have requested a feature that sends these details to a printer to be able to take them to their library on a sheet of paper.

## How to do this

We recommend forking this repository and creating a special git branch for this issue before you start. This will make it easier for you to submit a pull request later. If you've never done this before, there are a lot of good guides out there and you can always ask us questions.

### Adding the print function
In this part, you are going to add a "Print" button to the app that prints a sample HTML document.

* Have a look at the [Documentation](https://developer.android.com/training/printing/html-docs.html) on printing HTML documents to familiarize yourself with the API you are going to use.
* In the [menu resource file](https://github.com/opacapp/opacclient/blob/master/opacclient/opacapp/src/main/res/menu/search_result_details_activity.xml), add a new menu item labelled "Print". If you like, you can also add an appropriate Material Design icon and set it to show as an action button if there is room.
* In the `onOptionsItemSelected(MenuItem)` method of [SearchResultDetailFragment.java](https://github.com/opacapp/opacclient/blob/master/opacclient/opacapp/src/main/java/de/geeksfactory/opacclient/frontend/SearchResultDetailFragment.java#L577), add an additional case to handle clicking the "Print" button. You can create a new method that is called when this button is clicked.
* Add some code to your method to print a sample HTML document when the button is clicked, similar to the sample in the documentation mentioned above.
* Make sure that your code is only called on Android 4.4 (API level 19) and above and hide the "Print" button on other versions. The Printing API is not available on lower versions.
* Make sure that the details are already loaded when the print button is clicked, similarly to how it is done for the [Favorite button](https://github.com/opacapp/opacclient/blob/master/opacclient/opacapp/src/main/java/de/geeksfactory/opacclient/frontend/SearchResultDetailFragment.java#L693).
* Test the printing function by running the app on your device or an Android Emulator.

### Generating the HTML output
In this part, you are going to write the code that generates the actual HTML that is going to be printed, using the data from the detailled search result. This can be done conveniently using a templating language such as _mustache_.

* Have a look at some [documentation](https://mustache.github.io/mustache.5.html) about the _mustache_ templating language and its [usage](https://github.com/spullara/mustache.java/blob/master/README.md) in Java.
* Add the _mustache.java_ library to the project by inserting the following line into the [build.gradle](https://github.com/opacapp/opacclient/blob/master/opacclient/opacapp/build.gradle) file:
```gradle
    compile 'com.github.spullara.mustache.java:compiler:0.8.18'
```
* We have prepared a [sample](https://drop.rami.io/GHkmIX/) how the generated HTML should look like. Copy this HTML code to a new `.mustache` file in the `/src/main/res/raw` folder and adjust it so that it contains _mustache_ templating tags instead of the sample data. The documentation linked above should be helpful for this.
* In the [utils](https://github.com/opacapp/opacclient/tree/master/opacclient/opacapp/src/main/java/de/geeksfactory/opacclient/utils) package, create a new class called `PrintUtils` that contains a static method called `printDetail`. This should take a [DetailledItem](https://github.com/opacapp/opacclient/blob/master/opacclient/libopac/src/main/java/de/geeksfactory/opacclient/objects/DetailledItem.java) as a parameter and return HTML code as a string.
* Implement this method by using the _mustache.java_ library, similar to this:
```java
    InputStream is = getContext().getResources().openRawResource(resourceId);
    BufferedReader r = new BufferedReader(new InputStreamReader(is));

    MustacheFactory mf = new DefaultMustacheFactory();
    Mustache mustache = mf.compile(r, "print_template.mustache");
    StringWriter sw = new StringWriter();
    mustache.execute(sw, detailledItem).flush();
    return sw.toString();
```
* Go back to SearchResultDetailFragment.java and use the new `printDetail` function to generate the HTML from the `DetailledItem` stored in the `item` field.
* Test the printing feature once again.

We're excited to see your pull request!

If you get stuck, don't hestitate to ask. We bound ourselves to a [Code of Conduct](https://github.com/opacapp/opacclient/blob/master/CODE_OF_CONDUCT.md) and we're committed to providing a welcoming experience and giving constructive, friendly and useful feedback on your contributions.

**Comment by ceasarj** [comment]
Hello, if nobody is working on this issue the I would love to work on it.

**Comment by raphaelm** [comment]
Hey @ceasarj, we don't know of anyone else. Feel free to give it a go and just ask us if you get stuck!

**Comment by johan12345** [comment]
Fixed with #440.
